man/man1/geod.1
.\" @(#)geod.1
.nr LL 7.0i
.ad b
.hy 1
.TH GEOD 1 "2016/02/16 Rel. 4.9.3"
.SH NAME
geod \- direct geodesic computations
.br
invgeod \- inverse geodesic computations
.SH SYNOPSIS
.B geod
.B +ellps=<ellipse>
[
.B \-afFIlptwW
[
.I args
] ] [
.B +args
]
file[s]
.br
.B invgeod
.B +ellps=<ellipse>
[
.B \-afFIlptwW
[
.I args
] ] [
.B +args
]
file[s]
.SH DESCRIPTION
.I geod
(direct) and
.I invgeod
(inverse)
perform geodesic ("Great Circle") computations for determining
latitude, longitude and back azimuth of a terminus point
given a initial point latitude, longitude, azimuth and distance (direct) or
the forward and back azimuths and distance between an initial and
terminus point latitudes and longitudes (inverse).  The results are
accurate to round off for |\fIf\fR| < 1/50, where \fIf\fR is flattening.
.B invgeod
may not be available on all platforms; in this case call
.B geod
with the
.B \-I
option.
.PP
The following command-line options can appear in any order:
.TP
.B \-I
Specifies that the inverse geodesic computation is to be performed.
May be used with execution of
.B geod
as an alternative to
.B invgeod
execution.
.TP
.B \-a
Latitude and longitudes of the initial and terminal points,
forward and back azimuths and distance are output.
.TP
.BI \-t "a"
.I A
specifies a character employed as the first character to denote
a control line to be passed through without processing.
.TP
.BI \-le
Gives a listing of all the ellipsoids that may be selected with the
.B +ellps=
option.
.TP
.BI \-lu
Gives a listing of all the units that may be selected with the
.B +units=
option.
.TP
.BI \-[f|F] " format"
.I Format
is a
.I printf
format string to control the output form of the geographic coordinate values
(\fBf\fR) or distance value (\fBF\fR).
The default mode is DMS for geographic coordinates and "%.3f" for distance.
.TP
.BI \-[w|W] n
.I N
is the number of significant fractional digits to employ for
seconds output (when the option is not specified,
.B \-w3
is assumed).
When
.B \-W
is employed the fields will be constant width with leading zeroes.
.TP
.B \-p
This option causes the azimuthal values to be output as unsigned
DMS numbers between 0 and 360 degrees.  Also note \-f.
.PP
The
.B +args
command-line options are associated with geodetic parameters
for specifying the ellipsoidal or sphere to use.
See
.B proj
documentation for full list of these parameters and controls.
The options are processed in left to right order
from the command line.
Reentry of an option is ignored with the first occurrence assumed to
be the desired value.
.PP
One or more
.I files
(processed in left to right order)
specify the source of data to be transformed.
A \- will specify the location of processing standard input.
If no files are specified, the input is assumed to be from
.I stdin.
.PP
For direct determinations input data must be in latitude,
longitude, azimuth and distance order and output will be
latitude, longitude and back azimuth of the terminus point.
Latitude, longitude of the initial and terminus point are
input for the inverse mode and respective forward and back
azimuth from the initial and terminus points are output along
with the distance between the points.
.PP
Input geographic coordinates (latitude and longitude) and azimuthal data
must be in decimal degrees or DMS format and input distance data must be
in units consistent with the ellipsoid major axis or sphere radius
units.  The latitude must lie in the range [-90d,90d].  Output
geographic coordinates will be in DMS (if the
.B \-f
switch is not employed) to 0.001"
with trailing, zero-valued minute-second fields deleted.
Output distance data will be in the same units as the ellipsoid or
sphere radius.
.PP
The Earth's ellipsoidal figure may be selected in the same
manner as program
.B proj
by using
.B "+ellps=, +a=, +es=,"
etc.
.PP
.I Geod
may also be used to determine intermediate points along either
a geodesic line between two points or along an arc of specified distance
from a geographic point.
In both cases an initial point must be specified with
.BI +lat_1= lat
and
.BI +lon_1= lon
parameters and either a terminus point
.BI +lat_2= lat
and
.BI +lon_2= lon
or a distance and azimuth from the initial point with
.BI +S= distance
and
.BI +A= azimuth
must be specified.
.PP
If points along a geodesic are to be determined then either
.BI +n_S= integer
specifying the number of intermediate points and/or
.BI +del_S= distance
specifying the incremental distance between points must be specified.
.PP
To determine points along an arc equidistant from the initial point both
.BI +del_A= angle
and
.BI +n_A= integer
must be specified which determine the respective angular increments
and number of points to be determined.
.RE
.SH EXAMPLE
The following script determines the geodesic azimuths and distance in
U.S. statute miles from Boston, MA, to Portland, OR:
.RS 5
 \f(CWgeod +ellps=clrk66 <<EOF \-I +units=us\-mi
 42d15'N 71d07'W 45d31'N 123d41'W
 EOF\fR
.RE
which gives the results:
.RS 5
 \f(CW-66d31'50.141" 75d39'13.083" 2587.504
.RE
where the first two values are the
azimuth from Boston to Portland, the back azimuth from Portland to
Boston followed by the distance.
.PP
An example of forward geodesic use is to use the Boston location and determine
Portland's location by azimuth and distance:
.RS 5
 \f(CWgeod +ellps=clrk66 <<EOF +units=us-mi
 42d15'N 71d07'W \-66d31'50.141" 2587.504
 EOF\fR
.RE
which gives:
.RS 5
 \f(CW45d31'0.003"N 123d40'59.985"W 75d39'13.094"\fR
.RE
Note: lack of precision in the distance value compromises
the precision of the Portland location.
.SH SEE ALSO
.B geodesic(3)
.PP
\fBGeographicLib\fR, http://geographiclib.sf.net
.PP
The \fBGeodSolve\fR utility in GeographicLib.  With the \fB-E\fR option,
this solves the geodesic problems in terms of elliptic integrals; the
results are accurate for arbitrary \fIf\fR.
.PP
C. F. F. Karney, \fIAlgorithms for Geodesics\fR,
.br
J. Geodesy \fB87\fR, 43-55 (2013);
.br
DOI: http://dx.doi.org/10.1007/s00190-012-0578-z
.br
http://geographiclib.sf.net/geod-addenda.html
.PP
The \fIonline geodesic bibliography\fR,
.br
http://geographiclib.sf.net/geodesic-papers/biblio.html
.SH BUGS
A list of known bugs can found at https://github.com/OSGeo/proj.4/issues
where new bug reports can be submitted too.
.SH HOME PAGE
http://proj4.org/


NEWS
= GDAL/OGR 1.11.3 Release Notes = 

The 1.11.3 release is a bug fix release.

== Build (Unix) ==
 * Support custom buildflags in GRASS plugin (#5850)
 * GRASS 7.0.0 support (#5852)
 * fix detection of ECW SDK 5.1 on Mac (#5867, #5390)
 * Add support for Poppler >= 0.31.0 (#6118)

== Port ==
 * /vsitar/: remove useless validation test that prevents from opening valid .tar files (#5864)

== Algorithms ==
 * gdal_grid: fix crash in optimized mode with GCC 4.4 on 64bit (#5987)
 * Fix failure in GDALSuggestedWarpOut2() when computing output image size in case the top-left and bottom-right corners tranform to the same point (#5980)
 * Rasterize: add check to avoid burning negative x values (#5641)

== GDAL core ==
 * GMLJP2 reader: add compatibility with OGC CRS URL as found in GMLJP2v2 (#5940)
 * GMLJP2: on reading, don't do axis inversation if there's an explicit axisName requesting easting, northing order (#5960)
 * GMLJP2: add missing rangeParameters element to validate against GMLJP2 schema (#5707)

== Utilities ==
 * gdalbuildvrt: fix potential crash when using -b switch (#6095)
 * gdalserver: fix compilation with recent GNU libc (such as in Arch Linux) (#6073)
 * gdalwarp: emit error message if file specified with -cutline cannot be opened (#5921)
 * gdalwarp: fix 1.11.2 regression when invoking several times gdalwarp with several input files on the same target file, and when the input files have a nodata setting: only the last input file was warped onto the target dataset (#5909)
 * ogr2ogr: fix crash with -clipdst when a reprojection fails before (#5973)
 * ogr_layer_algebra.py: for Update, Clip and Erase, only creates attribute of input layer by default (#5976)

== GDAL drivers ==

ECRG driver:
 * change subdataset definition to make sure that they only consist of frames of same scale (#6043)

ERS driver:
 * fix SetProjection() (#5840)

GeoRaster driver:
 * New SRID search (#5911)
 * Fix user-defined SRID issue (#5881)

GTiff driver:
 * fix GTiffDataset::IsBlockAvailable() wrong behaviour when compiling against internal libtiff, when a BigTIFF file has a TileByteCounts with LONG/4-byte counts and not LONG8 (#6001)
 * avoid generated corrupted  right-most and bottom-most tiles for 12-bit JPEG-compressed (#5971)
 * on reading attach PROJ4 extension node for Mercator_Auxiliary_Sphere / WebMercator (#5924)

HDF4 driver:
 * fix AnyTypeToDouble() to use proper type (int instead of long) to work with DFNT_INT32/DFNT_UINT32 on 64-bit Linux (#5965)

HF2 driver:
 * fix reading side of the driver to work on architectures where char is unsigned, like PPC or ARM (#6082)

NetCDF driver:
 * fix crash on opening a NOAA dataset (#5962)
 * fix computation of inverse flattening (#5858)

Northwoord driver:
 * Fix computation of intermediate color table values on non-Intel platforms (#6091)
 * NWT_GRD: don't advertize scale/offset as they are transparently applied in IReadBlock() (#5839).

RPFTOC driver:
 * add tweak for weird relative directory names in the A.TOC file (#5979)

TIL driver:
 * fix half pixel shift in geo registration (#5961)

VRT driver:
 * add more checks to CheckCompatibleForDatasetIO() to avoid issues with overview bands (#5954)
 * fix rounding of output window size on VRTSimpleSource (#5874)

== OGRSpatialReference ==
 * morphFromESRI(): map Mercator_Auxiliary_Sphere to EPSG:3857 (#5924)

== OGR drivers ==

EDIGEO driver:
 * backport conversion of atof() to CPLAtof() to avoid truncation of floating point numbers including coordinates on a non C locale (#6070)

GML driver:
 * accept choice between gml:polygonProperty and gml:multiPolygonProperty (#5977)
 * fix GML_ATTRIBUTES_TO_OGR_FIELDS=YES to work correctly with xlink:href too (#5970); and also backport a trunk fix for writing the correctly path to attributes in nested elements
 * VFR: add support for OriginalniHraniceOmpv (#5853)

GeoJSON driver:
 * avoid truncation of real numbers on reading (#5882)
 * TopoJSON: fix segfault when reading https://raw.githubusercontent.com/mbostock/topojson/master/test/topojson/properties-id-computed.json (#5919)
 * GeoJSON writer: make string comparison for authority name case insensitive so as to recognize lowercase 'epsg' (#4995)

GPKG driver:
 * on reading, recognize GEOMCOLLECTION to be compatible of GDAL 2.0. Still writing GEOMETRYCOLLECTION in 1.11 branch (#5937)
 * escape column names on reading/insert/update (#5879)

MSSQLSpatial driver:
 * Add support for creating features with default values (#5957)

OpenFileGDB driver:
 * fix 1.11.2 regression with FileGDB v9 where the presence of a non spatial table can cause other layers not to be listed (#5875)

PG/PGDump drivers:
 * fix truncation of fields to work with multi-byte UTF-8 characters (#5854)

Shape driver:
 * SHPReadOGRFeature() should free the passed psShape when not NULL and if the DBF record is deleted, so as to avoid later reads to faild (linked to #5974)

SQLite driver:
 * Fix issue with consecutive identical characters in layer name with ExecuteSQL() (#6107)

== SWIG Language Bindings ==

Perl bindings:
 * Fix creation of Field Definition object (#6099)

Python bindings:
 * Fix build with SWIG 3.0.6 (#6045)

= GDAL/OGR 1.11.2 Release Notes = 

The 1.11.2 release is a bug fix release.

== Build ==

  * configure: fix detection of OCI by changing linking order to please modern GCC (#5550)
  * configure: fix test to accept MariaDB 10.X as valid MySQL (#5722)
  * More compiler warnings addressed (#5414)
  * Windows build of PDF: fix compilation issue with Visual Studio 2012 (#5744)
  * Windows build: Add support for MrSID 9.1 SDK (#5814)
  * Windows build: when building netCDF, HDF4, HDF5 as plugins, call registration of 'sub-drivers' GMT, HDF4Image and HDF5Image (#5802)

== Port ==

 * CSV finder: Stop probing for csv/horiz_cs.csv. (#5698)
 * /vsicurl/: avoid reading after end-of-file and fix failure when reading more than 16MB in a single time (#5786)
 * CPLHexToBinary(): faster implementation (#5812)


src/geod.c
/* <<<< Geodesic filter program >>>> */
# include "projects.h"
# include "geod_interface.h"
# include "emess.h"
# include <ctype.h>
# include <stdio.h>
# include <string.h>

# define MAXLINE 200
# define MAX_PARGS 50
# define TAB putchar('\t')
	static int
fullout = 0,	/* output full set of geodesic values */
tag = '#',	/* beginning of line tag character */
pos_azi = 0,	/* output azimuths as positive values */
inverse = 0;	/* != 0 then inverse geodesic */
	static char
*oform = (char *)0,	/* output format for decimal degrees */
*osform = "%.3f",	/* output format for S */
pline[50],		/* work string */
*usage =
"%s\nusage: %s [ -afFIptTwW [args] ] [ +opts[=arg] ] [ files ]\n";
	static void
printLL(double p, double l) {
	if (oform) {
		(void)printf(oform, p * RAD_TO_DEG); TAB;
		(void)printf(oform, l * RAD_TO_DEG);
	} else {
		(void)fputs(rtodms(pline, p, 'N', 'S'),stdout); TAB;
		(void)fputs(rtodms(pline, l, 'E', 'W'),stdout);
	}
}
	static void
do_arc(void) {
	double az;

	printLL(phi2, lam2); putchar('\n');
	for (az = al12; n_alpha--; ) {
		al12 = az = adjlon(az + del_alpha);
		geod_pre();
		geod_for();
		printLL(phi2, lam2); putchar('\n');
	}
}
	static void	/* generate intermediate geodesic coordinates */
do_geod(void) {
	double phil, laml, del_S;

	phil = phi2;
	laml = lam2;
	printLL(phi1, lam1); putchar('\n');
	for ( geod_S = del_S = geod_S / n_S; --n_S; geod_S += del_S) {
		geod_for();
		printLL(phi2, lam2); putchar('\n');
	}
	printLL(phil, laml); putchar('\n');
}
	static void	/* file processing function */
process(FILE *fid) {
	char line[MAXLINE+3], *s;

	for (;;) {
		++emess_dat.File_line;
		if (!(s = fgets(line, MAXLINE, fid)))
			break;
		if (!strchr(s, '\n')) { /* overlong line */
			int c;
			strcat(s, "\n");
			/* gobble up to newline */
			while ((c = fgetc(fid)) != EOF && c != '\n') ;
		}
		if (*s == tag) {
			fputs(line, stdout);
			continue;
		}
		phi1 = dmstor(s, &s);
		lam1 = dmstor(s, &s);
		if (inverse) {
			phi2 = dmstor(s, &s);
			lam2 = dmstor(s, &s);
			geod_inv();
		} else {
			al12 = dmstor(s, &s);
			geod_S = strtod(s, &s) * to_meter;
			geod_pre();
			geod_for();
		}
		if (!*s && (s > line)) --s; /* assumed we gobbled \n */
		if (pos_azi) {
			if (al12 < 0.) al12 += M_TWOPI;
			if (al21 < 0.) al21 += M_TWOPI;
		}
		if (fullout) {
			printLL(phi1, lam1); TAB;
			printLL(phi2, lam2); TAB;
			if (oform) {
				(void)printf(oform, al12 * RAD_TO_DEG); TAB;
				(void)printf(oform, al21 * RAD_TO_DEG); TAB;
				(void)printf(osform, geod_S * fr_meter);
			}  else {
				(void)fputs(rtodms(pline, al12, 0, 0), stdout); TAB;
				(void)fputs(rtodms(pline, al21, 0, 0), stdout); TAB;
				(void)printf(osform, geod_S * fr_meter);
			}
		} else if (inverse)
			if (oform) {
				(void)printf(oform, al12 * RAD_TO_DEG); TAB;
				(void)printf(oform, al21 * RAD_TO_DEG); TAB;
				(void)printf(osform, geod_S * fr_meter);
			} else {
				(void)fputs(rtodms(pline, al12, 0, 0), stdout); TAB;
				(void)fputs(rtodms(pline, al21, 0, 0), stdout); TAB;
				(void)printf(osform, geod_S * fr_meter);
			}
		else {
			printLL(phi2, lam2); TAB;
			if (oform)
				(void)printf(oform, al21 * RAD_TO_DEG);
			else
				(void)fputs(rtodms(pline, al21, 0, 0), stdout);
		}
		(void)fputs(s, stdout);
	}
}

static char *pargv[MAX_PARGS];
static int   pargc = 0;

int main(int argc, char **argv) {
	char *arg, **eargv = argv, *strnchr();
	FILE *fid;
	static int eargc = 0, c;

	if ((emess_dat.Prog_name = strrchr(*argv,'/')) != NULL) ++emess_dat.Prog_name;
	else emess_dat.Prog_name = *argv;
	inverse = ! strncmp(emess_dat.Prog_name, "inv", 3);
	if (argc <= 1 ) {
		(void)fprintf(stderr, usage, pj_get_release(),
                              emess_dat.Prog_name);
		exit (0);
	}
		/* process run line arguments */
	while (--argc > 0) { /* collect run line arguments */
		if(**++argv == '-') for(arg = *argv;;) {
			switch(*++arg) {
			case '\0': /* position of "stdin" */
				if (arg[-1] == '-') eargv[eargc++] = "-";
				break;
			case 'a': /* output full set of values */
				fullout = 1;
				continue;
			case 'I': /* alt. inverse spec. */
				inverse = 1;
				continue;
			case 't': /* set col. one char */
				if (arg[1]) tag = *++arg;
				else emess(1,"missing -t col. 1 tag");
				continue;
			case 'W': /* specify seconds precision */
			case 'w': /* -W for constant field width */
				if ((c = arg[1]) && isdigit(c)) {
					set_rtodms(c - '0', *arg == 'W');
					++arg;
				} else
				    emess(1,"-W argument missing or non-digit");
				continue;
			case 'f': /* alternate output format degrees or xy */
				if (--argc <= 0)
noargument:		   emess(1,"missing argument for -%c",*arg);
				oform = *++argv;
				continue;
			case 'F': /* alternate output format degrees or xy */
				if (--argc <= 0) goto noargument;
				osform = *++argv;
				continue;
			case 'l':
				if (!arg[1] || arg[1] == 'e') { /* list of ellipsoids */
                                    struct PJ_ELLPS *le;
                                    
                                    for (le=pj_get_ellps_ref(); le->id ; ++le)
                                        (void)printf("%9s %-16s %-16s %s\n",
                                                     le->id, le->major, le->ell, le->name);
				} else if (arg[1] == 'u') { /* list of units */
                                    struct PJ_UNITS *lu;
                                    
                                    for (lu = pj_get_units_ref();lu->id ; ++lu)
                                        (void)printf("%12s %-20s %s\n",
                                                     lu->id, lu->to_meter, lu->name);
				} else
                                    emess(1,"invalid list option: l%c",arg[1]);
                                exit( 0 );
			case 'p': /* output azimuths as positive */
				pos_azi = 1;
				continue;
			default:
				emess(1, "invalid option: -%c",*arg);
				break;
			}
			break;
		} else if (**argv == '+') /* + argument */
			if (pargc < MAX_PARGS)
				pargv[pargc++] = *argv + 1;
			else
				emess(1,"overflowed + argument table");
		else /* assumed to be input file name(s) */
			eargv[eargc++] = *argv;
	}
	/* done with parameter and control input */
	geod_set(pargc, pargv); /* setup projection */
	if ((n_alpha || n_S) && eargc)
		emess(1,"files specified for arc/geodesic mode");
	if (n_alpha)
		do_arc();
	else if (n_S)
		do_geod();
	else { /* process input file list */
		if (eargc == 0) /* if no specific files force sysin */
			eargv[eargc++] = "-";
		for ( ; eargc-- ; ++eargv) {
			if (**eargv == '-') {
				fid = stdin;
				emess_dat.File_name = "<stdin>";
			} else {
				if ((fid = fopen(*eargv, "r")) == NULL) {
					emess(-2, *eargv, "input file");
					continue;
				}
				emess_dat.File_name = *eargv;
			}
			emess_dat.File_line = 0;
			process(fid);
			(void)fclose(fid);
			emess_dat.File_name = (char *)0;
		}
	}
	exit(0); /* normal completion */
}
